p2p/addrbook.go

```diff
@@ -201,40 +210,34 @@ func (a *AddrBook) PickAddress(newBias int) *NetAddress {
 	oldCorrelation := math.Sqrt(float64(a.nOld)) * (100.0 - float64(newBias))
 	newCorrelation := math.Sqrt(float64(a.nNew)) * float64(newBias)
 
-	if (newCorrelation+oldCorrelation)*a.rand.Float64() < oldCorrelation {
-		// pick random Old bucket.
-		var bucket map[string]*knownAddress = nil
-		for len(bucket) == 0 {
-			bucket = a.addrOld[a.rand.Intn(len(a.addrOld))]
-		}
-		// pick a random ka from bucket.
-		randIndex := a.rand.Intn(len(bucket))
-		for _, ka := range bucket {
-			if randIndex == 0 {
-				return ka.Addr
-			}
-			randIndex--
-		}
-		cmn.PanicSanity("Should not happen")
-	} else {
-		// pick random New bucket.
-		var bucket map[string]*knownAddress = nil
-		for len(bucket) == 0 {
-			bucket = a.addrNew[a.rand.Intn(len(a.addrNew))]
+	// pick a random peer from a random bucket
+	var bucket map[string]*knownAddress
+	pickFromOldBucket := (newCorrelation+oldCorrelation)*a.rand.Float64() < oldCorrelation
+	if (pickFromOldBucket && a.nOld == 0) ||
+		(!pickFromOldBucket && a.nNew == 0) {
+		return nil
```

Comment:
Isn't this a behavioral change? The previous code never returned a nil address. It panicked in the worst case.

Reply:
This fixes a bug! Used to be we could get stuck in the infinite loop otherwise. And this function returns nil if the book is empty already, so this is another case when we return nil. I'll update the comment - thanks for the keen eye.

Reply:
Awesome, dude the new comments are great!